src/Coordination/KeeperStorage.cpp

```diff
@@ -197,362 +203,778 @@ void KeeperStorage::Node::removeChild(StringRef child_path)
     children.erase(child_path);
 }
 
-KeeperStorage::KeeperStorage(int64_t tick_time_ms, const String & superdigest_)
-    : session_expiry_queue(tick_time_ms)
-    , superdigest(superdigest_)
+void KeeperStorage::Node::invalidateDigestCache() const
+{
+    cached_digest.reset();
+}
+
+UInt64 KeeperStorage::Node::getDigest(const std::string_view path) const
+{
+    if (!cached_digest)
+        cached_digest = calculateDigest(path, data, stat);
+
+    return *cached_digest;
+};
+
+void KeeperStorage::Node::shallowCopy(const KeeperStorage::Node & other)
 {
-    container.insert("/", Node());
+    stat = other.stat;
+    seq_num = other.seq_num;
+    setData(other.getData());
```

Comment:
no, it's shallow because we copy only subset of data we need for preprocessing and digest calculation (e.g. we don't need to know the children of a node).
Will add a comment.